helix-core/src/main/java/org/apache/helix/model/Message.java

```diff
@@ -953,4 +970,35 @@ public boolean isValid() {
     }
     return true;
   }
+
+  // A class represents session aware ZNRecord for message. The message should be written to zk
+  // by the expected session.
+  // TODO: remove this class once public session-aware ZNRecord is available
+  private static class SessionAwareZNRecord extends ZNRecord implements SessionAwareZkWriteData {
```

Comment:
I still do not think adding this new subtype ZNRecord is an elegant solution here, specially you put it as a private class here only for message could make it look a bit hacky. Let us discuss what other options here.

Reply:
This is the most challenging part remaining in the PR that we consider. Multiple options are considered and tried. This one is by far a relatively cleaner one.

Considering:
1. not adding new session aware methods/interfaces ---> attach expected session together with data object
2. making it generic in zkclient ---> new interface `SessionAwareZkDataRecord`
3. Less hacky way than checking ZNRecord in zkclient ----> SessionAwareZNRecord extends ZNRecord implements SessionAwareZkWriteData. We don't have a public implementation `SessionAwareZNRecord`. So this is needed. I've added TODO if in future we implement public `SessionAwareZNRecord`, we should remove this private class.

Reply:
If we believe this is the right solution, why not make it public?

Reply:
We still want to make minimum changes in this PR. Putting it private gives us more flexibility to change in future: may add more stuff to this class. If in future we find that we need `SessionAwareZNRecord` in other places, we should create such thing as public. We need to discuss more and have a more comprehensive design for public `SessionAwareZNRecord`. What do you think?

Reply:
@lei-xia I've considered and tried to break the PR down to 2 smaller ones: fixing the issue without new interface and adding new interface. But the first one PR would look quite hacky and I don't feel comfortable merging the commit to code base. So I decided to keep this PR and make `SessionAwareZNRecord` public. Even in the future we need to add more APIs to the interface, we could also add API implementation to `SessionAwareZNRecord`. In this way, I feel it is cleaner to check in.